File: batch-tech-feed/produce/techfeed/media.py
"""ffmpeg / ffprobe の薄いラッパ。

映像は Remotion が焼くので、ここが受け持つのは音声と HLS 変換だけである。

音声の尺は「計算」せず必ず ffprobe で実測する(docs/407 §2.3)。TTS エンジンが返す
理論値とファイルの実尺は一致しないことがあり、その差が動画後半での絵・字幕の
ずれとして蓄積するため。
"""

from __future__ import annotations

import json
import subprocess

# 合成した各wavはここに揃える。エンジンやモックで出力形式が違っても、
# 後段の concat が「全ファイル同一フォーマット」を前提にできるようにするため。
SAMPLE_RATE = 48000
CHANNELS = 1


class MediaError(RuntimeError):
    pass


def _run(cmd: list[str]) -> subprocess.CompletedProcess:
    proc = subprocess.run(cmd, capture_output=True, text=True)
    if proc.returncode != 0:
        raise MediaError(f"{cmd[0]} failed ({proc.returncode}): {' '.join(cmd)}\n{proc.stderr[-2000:]}")
    return proc


def run_ffmpeg(args: list[str]) -> None:
    """任意の ffmpeg 呼び出し。共通のフラグだけこちらで付ける。"""
    _run(["ffmpeg", "-y", "-loglevel", "error", *args])


def probe_duration_ms(path: str) -> int:
    proc = _run(
        [
            "ffprobe", "-v", "error",
            "-show_entries", "format=duration",
            "-of", "json",
            path,
        ]
    )
    duration = json.loads(proc.stdout)["format"]["duration"]
    return int(round(float(duration) * 1000))


def normalize_audio(src: str, dst: str, pad_ms: int) -> None:
    """wavを共通フォーマットへ揃え、末尾に無音を足す。

    無音の付与はレンダリング後ではなくここで行う。文と文の「間」も尺の一部として
    実測されるべきで、後から足すとタイムラインとファイルが食い違うため。
    """
    filters = [f"aresample={SAMPLE_RATE}", f"aformat=channel_layouts=mono"]
    if pad_ms > 0:
        filters.append(f"apad=pad_dur={pad_ms / 1000:.3f}")

    _run(
        [
            "ffmpeg", "-y", "-loglevel", "error",
            "-i", src,
            "-af", ",".join(filters),
            "-ar", str(SAMPLE_RATE), "-ac", str(CHANNELS), "-c:a", "pcm_s16le",
            dst,
        ]
    )


def silence(dst: str, duration_ms: int) -> None:
    """指定尺の無音wavを作る。TTSエンジンなしでパイプラインを通すため(--engine mock)。"""
    _run(
        [
            "ffmpeg", "-y", "-loglevel", "error",
            "-f", "lavfi",
            "-i", f"anullsrc=r={SAMPLE_RATE}:cl=mono",
            "-t", f"{duration_ms / 1000:.3f}",
            "-c:a", "pcm_s16le",
            dst,
        ]
    )


def concat_audio(list_file: str, dst: str) -> None:
    _run(
        [
            "ffmpeg", "-y", "-loglevel", "error",
            "-f", "concat", "-safe", "0",
            "-i", list_file,
            "-c:a", "pcm_s16le",
            dst,
        ]
    )

File: batch-tech-feed/produce/shared/media.go
package shared

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"os/exec"
	"strconv"
	"strings"
)

const (
	SampleRate = 48000
	Channels   = 1
)

func runCmd(cmdName string, args ...string) (string, error) {
	cmd := exec.Command(cmdName, args...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		return "", fmt.Errorf("%s failed (%w):\n%s", cmdName, err, stderr.String())
	}
	return stdout.String(), nil
}

// RunFFmpeg は ffmpeg コマンドを実行する。
func RunFFmpeg(args ...string) error {
	fullArgs := append([]string{"-y", "-loglevel", "error"}, args...)
	_, err := runCmd("ffmpeg", fullArgs...)
	return err
}

// ProbeDurationMs は ffprobe で音声・動画ファイルの実測尺 (ms) を取得する。
func ProbeDurationMs(path string) (int, error) {
	args := []string{
		"-v", "error",
		"-show_entries", "format=duration",
		"-of", "json",
		path,
	}
	out, err := runCmd("ffprobe", args...)
	if err != nil {
		return 0, fmt.Errorf("ffprobe failed for %s: %w", path, err)
	}

	var parsed struct {
		Format struct {
			Duration string `json:"duration"`
		} `json:"format"`
	}
	if err := json.Unmarshal([]byte(out), &parsed); err != nil {
		return 0, fmt.Errorf("failed to parse ffprobe output: %w", err)
	}

	durFloat, err := strconv.ParseFloat(parsed.Format.Duration, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid duration format '%s': %w", parsed.Format.Duration, err)
	}

	return int(math.Round(durFloat * 1000.0)), nil
}

// NormalizeAudio は wav を共通フォーマット (48kHz 1ch PCM 16bit) へ揃え、末尾に指定パディングを付与する。
func NormalizeAudio(src, dst string, padMs int) error {
	filters := []string{
		fmt.Sprintf("aresample=%d", SampleRate),
		"aformat=channel_layouts=mono",
	}
	if padMs > 0 {
		filters = append(filters, fmt.Sprintf("apad=pad_dur=%.3f", float64(padMs)/1000.0))
	}

	args := []string{
		"-i", src,
		"-af", strings.Join(filters, ","),
		"-ar", strconv.Itoa(SampleRate),
		"-ac", strconv.Itoa(Channels),
		"-c:a", "pcm_s16le",
		dst,
	}
	return RunFFmpeg(args...)
}

// Silence は指定尺 (ms) の無音 wav ファイルを作成する。
func Silence(dst string, durationMs int) error {
	args := []string{
		"-f", "lavfi",
		"-i", fmt.Sprintf("anullsrc=r=%d:cl=mono", SampleRate),
		"-t", fmt.Sprintf("%.3f", float64(durationMs)/1000.0),
		"-c:a", "pcm_s16le",
		dst,
	}
	return RunFFmpeg(args...)
}

// ConcatAudio は listFile に記述された複数 wav ファイルを連結する。
func ConcatAudio(listFile, dst string) error {
	args := []string{
		"-f", "concat",
		"-safe", "0",
		"-i", listFile,
		"-c:a", "pcm_s16le",
		dst,
	}
	return RunFFmpeg(args...)
}
